Riot/Managers/Widgets/WidgetManager.m

```diff
@@ -450,65 +451,156 @@ - (MXHTTPOperation *)getScalarTokenForMXSession:(MXSession*)mxSession
-                                                                            failure(error);
-                                                                        }
-                                                                    }];
+         } failure:^(NSError *error) {
+             NSLog(@"[WidgetManager] registerForScalarToken. Error in modular/register request");
 
-                [operation mutateTo:operation2];
+             if (failure)
+             {
+                 failure(error);
+             }
+         }];
 
-            }
+        [operation mutateTo:operation2];
 
-        } failure:^(NSError *error) {
-            NSLog(@"[WidgetManager] getScalarTokenForMXSession. Error in openIdToken request");
+    } failure:^(NSError *error) {
+        NSLog(@"[WidgetManager] registerForScalarToken. Error in openIdToken request");
 
-            if (failure)
-            {
-                failure(error);
-            }
-        }];
-    }
+        if (failure)
+        {
+            failure(error);
+        }
+    }];
 
     return operation;
 }
 
+- (MXHTTPOperation *)validateScalarToken:(NSString*)scalarToken forMXSession:(MXSession*)mxSession
+                                complete:(void (^)(BOOL valid))complete
+                                 failure:(void (^)(NSError *error))failure
+{
+    NSString *modularRestUrl = [[NSUserDefaults standardUserDefaults] objectForKey:@"integrationsRestUrl"];
+    MXHTTPClient *httpClient = [[MXHTTPClient alloc] initWithBaseURL:modularRestUrl andOnUnrecognizedCertificateBlock:nil];
+
+    return [httpClient requestWithMethod:@"GET"
+                                    path:[NSString stringWithFormat:@"account?scalar_token=%@", scalarToken]
+                              parameters:nil
+                                 success:^(NSDictionary *JSONResponse) {
+
+                                     NSString *userId;
+                                     MXJSONModelSetString(userId, JSONResponse[@"user_id"])
+
+                                     if ([userId isEqualToString:mxSession.myUser.userId])
+                                     {
+                                         complete(YES);
+                                     }
+                                     else
+                                     {
+                                         NSLog(@"[WidgetManager] validateScalarToken. Unexpected modular/account response: %@", JSONResponse);
+                                         complete(NO);
+                                     }
+
+                                 } failure:^(NSError *error) {
+                                     NSHTTPURLResponse *urlResponse = [MXHTTPOperation urlResponseFromError:error];
+
+                                     NSLog(@"[WidgetManager] validateScalarToken. Error in modular/account request. statusCode: %@", @(urlResponse.statusCode));
+
+                                     if (urlResponse &&  urlResponse.statusCode / 100 != 2)
```

Comment:
It works but maybe `urlResponse.statusCode >= 200 & urlResponse.statusCode < 300` would be more readable

Reply:
This is very subjective.